### PR TITLE
[KYUUBI #3046][Metrics] Add meter metrics for recording the rate of the operation state for each kyuubi operation

### LINKED ISSUE #3046: [FEATURE] Expose the batch job submission metrics
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/incubator-kyuubi/issues?q=is%3Aissue) and found no similar issues.

### Describe the feature

Expose the batch session related metrics.

Such as:

```
pending batch sessions
running batch sessions

```

### Motivation

for server moniroting

### Describe the solution

_No response_

### Additional context

_No response_

### Are you willing to submit PR?

- [x] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by turboFei**
### _Why are the changes needed?_

Close #3046
Expose the metrics likes:

```
kyuubi.operation.state.BatchJobSubmission.pending
kyuubi.operation.state.BatchJobSubmission.running
kyuubi.operation.state.BatchJobSubmission.finished
kyuubi.operation.state.BatchJobSubmission.error
kyuubi.operation.state.BatchJobSubmission.canceled
```
So that the kyuubi service admin can know that how many BatchJobSubmission operations are pending, running and so on.

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by ulysses-you** [comment]
is it possible add some test ?

**Comment by turboFei** [comment]
> is it possible add some test ?

sure

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/JpsApplicationOperation.scala [hunk 1]
```diff
@@ -65,6 +65,9 @@ class JpsApplicationOperation extends ApplicationOperation {
       val (id, _) = idAndCmd.splitAt(idAndCmd.indexOf(" "))
       try {
         s"kill -15 $id".lineStream
+        if (getEngine(tag).nonEmpty) {
+          killApplicationByTag(tag)
```
Comment (turboFei) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/JpsApplicationOperation.scala:69]: to prevent flaky test.

Comment (turboFei) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/JpsApplicationOperation.scala:69]: spark-submit will generate two processes.

Comment (ulysses-you) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/JpsApplicationOperation.scala:69]: if so the first kill will fail since there is no that process, shall we check the result of the command ? if the process is not found we should try to kill again, otherwise do nothing

Comment (turboFei) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/JpsApplicationOperation.scala]: sounds good, I thought that the command will not throw exception even it failed to kill the process.

Just have a testing, it will throw exception if the command fail.

done.

**Comment by yaooqinn** [comment]
doc is missing

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/JpsApplicationOperation.scala [hunk 2]
```diff
@@ -68,7 +68,11 @@ class JpsApplicationOperation extends ApplicationOperation {
         (true, s"Succeeded to terminate: $idAndCmd")
       } catch {
         case e: Exception =>
-          (false, s"Failed to terminate: $idAndCmd, due to ${e.getMessage}")
+          if (getEngine(tag).nonEmpty) {
```
Comment (ulysses-you) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/JpsApplicationOperation.scala:71]: we should only try one time and fail if the last attempt does not succeed ?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/JpsApplicationOperation.scala [hunk 3]
```diff
@@ -68,7 +68,11 @@ class JpsApplicationOperation extends ApplicationOperation {
...
+            killApplicationByTag(tag)
+          } else {
+            (false, s"Failed to terminate: $idAndCmd, due to ${e.getMessage}")
+          }
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/JpsApplicationOperation.scala:75]: what is this for, seems not related to the scope of pr

Comment (ulysses-you) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/JpsApplicationOperation.scala:75]: yeah, we can fix this in a new pr

Comment (yaooqinn) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/JpsApplicationOperation.scala:75]: so, what are we going to fix?

Comment (turboFei) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/JpsApplicationOperation.scala:75]: the spark-submit may generate two process. 

And we might fail to kill the running batch jobs after each test in BatchResourceSuite.

Comment (turboFei) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/JpsApplicationOperation.scala:75]: It might cause that the new added UT failed.

Because the new UI assert that the running count is 0 at the begining.

Comment (yaooqinn) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/JpsApplicationOperation.scala:75]: let's not hide bugs...

File: kyuubi-server/src/main/scala/org/apache/kyuubi/operation/KyuubiOperation.scala [hunk 4]
```diff
@@ -156,8 +157,11 @@ abstract class KyuubiOperation(session: Session) extends AbstractOperation(sessi
   override def shouldRunAsync: Boolean = false
 
   override def setState(newState: OperationState): Unit = {
+    MetricsSystem.tracing { ms =>
+      ms.decCount(MetricRegistry.name(OPERATION_STATE, opType, state.toString.toLowerCase))
+      ms.incCount(MetricRegistry.name(OPERATION_STATE, opType, newState.toString.toLowerCase))
+      ms.markMeter(MetricRegistry.name(OPERATION_STATE, newState.toString.toLowerCase))
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/operation/KyuubiOperation.scala:163]: what is this for?

Comment (turboFei) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/operation/KyuubiOperation.scala:163]: I think we can remove this one, it seems meaning less and always increasing.

Comment (yaooqinn) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/operation/KyuubiOperation.scala:163]: what is different from L43

File: kyuubi-server/src/main/scala/org/apache/kyuubi/operation/KyuubiOperation.scala [hunk 5]
```diff
@@ -156,8 +157,11 @@ abstract class KyuubiOperation(session: Session) extends AbstractOperation(sessi
   override def shouldRunAsync: Boolean = false
 
   override def setState(newState: OperationState): Unit = {
+    MetricsSystem.tracing { ms =>
+      ms.decCount(MetricRegistry.name(OPERATION_STATE, opType, state.toString.toLowerCase))
+      ms.incCount(MetricRegistry.name(OPERATION_STATE, opType, newState.toString.toLowerCase))
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/operation/KyuubiOperation.scala:162]: so the terminal states are cumulative, but the intermediate ones are not

File: docs/monitor/metrics.md [hunk 6]
```diff
@@ -89,3 +89,12 @@ Since v1.5.0, you can use the following metrics to replace:
 - `kyuubi.operation.total.ExecuteStatement`
 - `kyuubi.operation.opened.ExecuteStatement`
 - `kyuubi.operation.failed.ExecuteStatement.${errorType}`
+
+Since v1.6.0, you can use the following metrics to get the current state counter for a specified operation type:
```
Comment (yaooqinn) [inline docs/monitor/metrics.md:93]: the doc is eligible. please follow the table above. the one you added can only be an extra hint.

File: docs/monitor/metrics.md [hunk 7]
```diff
@@ -79,6 +79,7 @@ Metrics Prefix | Metrics Suffix | Type | Since | Description
 `kyuubi.backend_service.fetch_result_rows_rate`  | | meter | 1.5.0 |<div style='width: 150pt;word-wrap: break-word;white-space: normal'> kyuubi backend service `fetchResults` method that fetch result rows rate </div>
 `kyuubi.backend_service.get_primary_keys`  | | meter | 1.6.0 |<div style='width: 150pt;word-wrap: break-word;white-space: normal'> kyuubi backend service `get_primary_keys` method execution time and rate </div>
 `kyuubi.backend_service.get_cross_reference`  | | meter | 1.6.0 |<div style='width: 150pt;word-wrap: break-word;white-space: normal'> kyuubi backend service `get_cross_reference` method execution time and rate </div>
+`kyuubi.operation.state`   | `${operationType}`<br/>`.${state}`     | counter | 1.6.0 |<div style='width: 150pt;word-wrap: break-word;white-space: normal'>  current count for the `${operationType}` with a particular `${state}`, e.g. `BatchJobSubmission.pending`</div>
```
Comment (yaooqinn) [inline docs/monitor/metrics.md:82]: you can use meters for this actually.. just like kyuubi.operation.state.$state

```
"kyuubi.operation.state.closed" : {
      "count" : 19,
      "m15_rate" : 0.020137396109391522,
      "m1_rate" : 1.6595792398582304E-4,
      "m5_rate" : 0.003290775624396568,
      "mean_rate" : 0.00539128492410111,
      "units" : "events/second"
    },
    "kyuubi.operation.state.error" : {
      "count" : 1,
      "m15_rate" : 0.025462486603774714,
      "m1_rate" : 7.483284619160424E-15,
      "m5_rate" : 4.1270758361306424E-4,
      "mean_rate" : 5.372806924825531E-4,
      "units" : "events/second"
    },
    "kyuubi.operation.state.finished" : {
      "count" : 18,
      "m15_rate" : 0.01815503925250486,
      "m1_rate" : 1.5388408901182632E-14,
      "m5_rate" : 1.1365883986004687E-4,
      "mean_rate" : 0.005107080000702363,
      "units" : "events/second"
    },
    "kyuubi.operation.state.initialized" : {
      "count" : 19,
      "m15_rate" : 0.006237933279653059,
      "m1_rate" : 1.5785160613558252E-14,
      "m5_rate" : 1.1526188920321656E-4,
      "mean_rate" : 0.005376601901198222,
      "units" : "events/second"
    }
```

File: kyuubi-server/src/main/scala/org/apache/kyuubi/operation/KyuubiOperation.scala (hunk 5 above)
Comment (turboFei) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/operation/KyuubiOperation.scala:162]: yes

File: kyuubi-server/src/main/scala/org/apache/kyuubi/operation/KyuubiOperation.scala (hunk 4 above)
Comment (turboFei) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/operation/KyuubiOperation.scala:163]: the state value is different with that of L43.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/operation/KyuubiOperation.scala (hunk 5 above)
Comment (yaooqinn) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/operation/KyuubiOperation.scala:162]: if so, you need to document it clear

File: docs/monitor/metrics.md (hunk 7 above)
Comment (turboFei) [reply docs/monitor/metrics.md:82]: thanks, let me have a try

File: kyuubi-server/src/main/scala/org/apache/kyuubi/operation/KyuubiOperation.scala (hunk 4 above)
Comment (yaooqinn) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/operation/KyuubiOperation.scala:163]: i see

File: docs/monitor/metrics.md (hunk 7 above)
Comment (turboFei) [reply docs/monitor/metrics.md:82]: thanks, it helps a lot.

Comment (yaooqinn) [reply docs/monitor/metrics.md:82]: the discussion is marked as resolved but the code part remains unchanged？

Comment (turboFei) [reply docs/monitor/metrics.md]: ...resolved locally.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/JpsApplicationOperation.scala (hunk 3 above)
Comment (turboFei) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/JpsApplicationOperation.scala]: have remove it from this pr.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/operation/BatchJobSubmission.scala [hunk 8]
```diff
@@ -306,9 +306,7 @@ class BatchJobSubmission(
         case e: IOException => error(e.getMessage, e)
       }
 
-      MetricsSystem.tracing(_.decCount(
-        MetricRegistry.name(OPERATION_OPEN, statement.toLowerCase(Locale.getDefault))))
```
Comment (ulysses-you) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/operation/BatchJobSubmission.scala:310]: shall we also remove the override of statement in this class ?

Comment (turboFei) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/operation/BatchJobSubmission.scala:310]: nice catch, removed

Comment (turboFei) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/operation/BatchJobSubmission.scala:310]: also remove the statement override in LaunchEngine: https://github.com/apache/incubator-kyuubi/pull/3077

**Review by yaooqinn — APPROVED** [review]

**Comment by turboFei** [comment]
thanks @yaooqinn @ulysses-you 

merging to master
